liquidator/test/Liquidator.js

```diff
@@ -430,6 +430,188 @@ contract("Liquidator.js", function(accounts) {
+      // A price of 25 USD per token will make all positions undercollateralized.
+      // Numerically debt * price * coltReq > debt for collateralized position.
+      // Sponsor1: 12 * 25 * 1.2 > 100
+      // Sponsor2: 8 * 25 * 1.2 > 100
+      // Sponsor3: 5 * 25 * 1.2 > 100
+      priceFeedMock.setCurrentPrice(toBN(toWei("25")));
+
+      await liquidator.queryAndLiquidate(amountToLiquidate);
+      assert.equal(spy.callCount, 3); // 2 info + 1 error level events should be sent at the conclusion of the 2 successful and 1 failed liquidations.
+      assert.equal(lastSpyLogLevel(spy), "error");
+      assert.equal(spy.getCall(-1).lastArg.tokensToLiquidate, "0");
+
+      // Sponsor1 should be in a liquidation state with the bot as the liquidator. (4/12) = 33.33% of the 100 starting collateral and 6 tokens should be liquidated.
+      let liquidationObject = (await emp.getLiquidations(sponsor1))[0];
+      assert.equal(liquidationObject.sponsor, sponsor1);
+      assert.equal(liquidationObject.liquidator, liquidatorBot);
+      assert.equal(liquidationObject.state, LiquidationStatesEnum.PRE_DISPUTE);
+      assert.equal(liquidationObject.liquidatedCollateral, toWei("33.3333333333333333"));
+      assert.equal(liquidationObject.tokensOutstanding, toWei("4"));
+
+      // Sponsor2 should be in a liquidation state with the bot as the liquidator. (3/8) = 37.5% of the 100 starting collateral and 3 tokens should be liquidated.
+      liquidationObject = (await emp.getLiquidations(sponsor2))[0];
+      assert.equal(liquidationObject.sponsor, sponsor2);
+      assert.equal(liquidationObject.liquidator, liquidatorBot);
+      assert.equal(liquidationObject.state, LiquidationStatesEnum.PRE_DISPUTE);
+      assert.equal(liquidationObject.liquidatedCollateral, toWei("37.5"));
+      assert.equal(liquidationObject.tokensOutstanding, toWei("3"));
+
+      // Sponsor3 should not have been liquidated.
+
+      // Sponsor1 should have some collateral and tokens left in their position from the liquidation.
+      assert.equal((await emp.getCollateral(sponsor1)).rawValue, toWei("66.6666666666666667"));
+      let positionObject = await emp.positions(sponsor1);
+      assert.equal(positionObject.tokensOutstanding.rawValue, toWei("8"));
+
+      // Sponsor2 should have some collateral and tokens left in their position from the liquidation.
+      assert.equal((await emp.getCollateral(sponsor2)).rawValue, toWei("62.5"));
+      positionObject = await emp.positions(sponsor2);
+      assert.equal(positionObject.tokensOutstanding.rawValue, toWei("5"));
+
+      // Sponsor3 should have its full position remaining.
+      assert.equal((await emp.getCollateral(sponsor3)).rawValue, toWei("100"));
+      positionObject = await emp.positions(sponsor3);
+      assert.equal(positionObject.tokensOutstanding.rawValue, toWei("5"));
+    });
+
+    it("bot does not have enough funds to liquidate all positions", async function() {
```

Comment:
@mrice32 the bot will fail silently when it doesn't have enough funds

Reply:
Does it emit an `error` level log? why is this silent?

Reply:
@nicholaspai I don't think we currently fail silently in that case. [This](https://github.com/UMAprotocol/protocol/blob/a7855f6a6333d60afb68a165090e6fb7f8801465/liquidator/liquidator.js#L218-L235) try catch sends out an `error` log, and I think that's where it would blow up previously if it didn't have enough inventory to liquidate. I think this is probably something we want to keep. I like that we'll do a partial liquidation now, but can we also send out an `error` log if we're forced to do that?

Reply:
Its an `error` log so its not "silent". I meant to say its silent because the bot doesn't crash, i.e. we intentionally catch this situation.

Reply:
There's actually an important assertion in this test that makes sure the log is an `error` log: `      assert.equal(lastSpyLogLevel(spy), "error");`